nncf/quantization/algorithms/bias_correction/algorithm.py

```diff
@@ -182,85 +190,84 @@ def _apply(
+    def _get_subgraph_data_for_node(self, node: NNCFNode, nncf_graph: NNCFGraph, model: TModel) -> Dict[str, List[str]]:
         """
         This method collects necessary data for the specified node and its subgraph.
         This data contains the nodes (NNCFNode) for the subgraph building
         and statistics collection (for the next step).
 
         :param node: NNCFNode instance. This is the main node that with bias that would be corrected (or not).
         :param nncf_graph: NNCFGraph instance for graph analysis.
+        :param model: Backend-specific model.
         :return: A dict with the list of the nodes for the subgraph input and statistics collection.
         """
-        stats_nodes, input_nodes, output_nodes = [], [], []
-
-        def traverse_to_layers_with_bias(node, output):
-            if node in output:
-                return True, output
-            if self._backend_entity.is_node_with_bias(node, nncf_graph):
-                output.append(node)
-                self._collected_stat_inputs.add(node.node_name)
-                activation_input = nncf_graph.get_input_edges(node)[0].from_node
-
-                output_nodes.append(activation_input)
-                return True, output
-            return False, output
-
-        def traverse_to_input_layers(node, output):
-            if node in output + input_nodes:
-                return True, output
-            if node.node_name in self._collected_stat_inputs and node not in stats_nodes:
-                output.append(node)
-                return True, output
-            return False, output
-
+        statistic_nodes, subgraph_input_nodes, subgraph_output_nodes = [], [], []
+
+        def fill_statistic_nodes(node):
+            # A small hack to speed up graph traversal.
+            if node in statistic_nodes or node in visited_nodes:
+                return
+            visited_nodes.append(node)
+
+            # If we found a node with bias, we have to collect it as a statistic node,
+            # and its input for _collected_stat_inputs_map,
+            # which will be used during the collection of statistics for the next node.
+            if self._backend_entity.is_quantized_weights(node, nncf_graph) and self._backend_entity.is_node_with_bias(
+                node, nncf_graph
+            ):
+                statistic_nodes.append(node)
+                activation_node, output_port_id = self._get_activation_node_and_port(node, nncf_graph)
+                subgraph_output_nodes.append(activation_node)
+                self._collected_stat_inputs_map[node.node_name] = (activation_node.node_name, output_port_id)
+                return
+
+            for next_node in nncf_graph.get_next_nodes(node):
+                fill_statistic_nodes(next_node)
+
+        def fill_subgraph_input_nodes(node):
+            # A small hack to speed up graph traversal.
+            if node in subgraph_input_nodes or node in visited_nodes:
+                return
```

Comment:
Is it possible that node is in `subgraph_input_nodes` and not in `visited_nodes`?